applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp

```diff
@@ -0,0 +1,743 @@
+template<class TSparseSpace,
+         class TDenseSpace,
+         class TLinearSolver
+         >
+class HarmonicAnalysisStrategy
+    : public SolvingStrategy<TSparseSpace, TDenseSpace, TLinearSolver>
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(HarmonicAnalysisStrategy);
+
+    typedef SolvingStrategy<TSparseSpace, TDenseSpace, TLinearSolver> BaseType;
+
+    typedef typename BaseType::TSchemeType::Pointer SchemePointerType;
+
+    typedef typename BaseType::TBuilderAndSolverType::Pointer BuilderAndSolverPointerType;
+
+    typedef typename TDenseSpace::VectorType DenseVectorType;
+
+    typedef typename TDenseSpace::MatrixType DenseMatrixType;
+
+    typedef typename TDenseSpace::MatrixPointerType DenseMatrixPointerType;
+
+    typedef TSparseSpace SparseSpaceType;
+
+    typedef typename TSparseSpace::VectorType SparseVectorType;
+    
+    typedef typename TSparseSpace::VectorPointerType SparseVectorPointerType;
+
+    typedef typename TSparseSpace::MatrixType SparseMatrixType;
+
+    typedef typename TSparseSpace::MatrixPointerType SparseMatrixPointerType;
+
+    typedef std::complex<double> ComplexType;
+
+    typedef boost::numeric::ublas::vector<ComplexType> ComplexVectorType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+    HarmonicAnalysisStrategy(
+        ModelPart& model_part,
+        SchemePointerType pScheme,
+        BuilderAndSolverPointerType pBuilderAndSolver,
+        bool UseMaterialDampingFlag = false
+        )
+        : SolvingStrategy<TSparseSpace, TDenseSpace, TLinearSolver>(model_part)
+    {
+        KRATOS_TRY
+
+        mpScheme = pScheme;
+
+        mpBuilderAndSolver = pBuilderAndSolver;
+
+        // ensure initialization of system matrices in InitializeSolutionStep()
+        mpBuilderAndSolver->SetDofSetIsInitializedFlag(false);
+
+        mInitializeWasPerformed = false;
+
+        SparseVectorType* AuxForceVector = new SparseVectorType;
+        mpForceVector = boost::shared_ptr<SparseVectorType>(AuxForceVector);
+
+        DenseMatrixType* AuxModalMatrix = new DenseMatrixType;
+        mpModalMatrix = boost::shared_ptr<DenseMatrixType>(AuxModalMatrix);
+
+        mRayleighAlpha = 0.0;
+        mRayleighBeta = 0.0;
+        mSystemDamping = 0.0;
+        this->SetUseMaterialDampingFlag(UseMaterialDampingFlag);
+
+        // default echo level (mute)
+        this->SetEchoLevel(0);
+
+        // default rebuild level (build only once)
+        this->SetRebuildLevel(0);
+
+        KRATOS_CATCH("")
+    }
+
+    /// Deleted copy constructor.
+    HarmonicAnalysisStrategy(const HarmonicAnalysisStrategy& Other) = delete;
+
+    /// Destructor.
+    virtual ~HarmonicAnalysisStrategy()
+    {
+        // Clear() controls order of deallocation to avoid invalid memory access
+        // in some special cases.
+        // warning: BaseType::GetModelPart() may be invalid here.
+        this->Clear();
+    }
```

Comment:
normally avoid throwing an exception in destructors. You could put in Finalize() for example.